content(schematic-helpers): add `externalSchematic` (#11)
d-koppenhagen/d-koppenhagen.de: blog/2020-09-angular-schematics-common-helpers.md

If you run a schematic, you may come to the point where one schematic should execute another one.
For example: You create Schematics for generating a specific component.
You also develop a `ng add` or `ng new` schematic to set up things for you and create an example component by default.
In such cases you may want to combine multiple Schematics.

> ### Run a local Schematic using the `RunSchematicTask`

Reviewer: "Run local schematics"

}
```

If we now run `schematics ../playground/src/collection.json:ng-add --debug=false` from our example Angular project, we can see that the `ng-add` schematic has called the `playground` schematic.

> With this knowledge you can define small atomic Schematics that can be executed 'standalone' or from another schematic that combines multiple standalone Schematics and calls them with specific parameters.

Reviewer: "schematics"

Author: Aber "Schematics" habe ich immer groß geschrieben?!

> With this knowledge you can define small atomic Schematics that can be executed 'standalone' or from another schematic that combines multiple standalone Schematics and calls them with specific parameters.

Reviewer: "standalone" with Double quotes?

Author: Arent't single quotes the right one in English language here? My spell checker plugin suggested single quotes also.

Reviewer: Piece from thre Angular docs:

> The exponentialStrength pipe executes every time the user changes the "normal power" value or the "boost factor", as shown in Figure 3.

https://angular.io/guide/pipes

- [Check out the example in the playground repository on GitHub](https://github.com/d-koppenhagen/schematics-helpers-playground/tree/master/playground/src/ng-add)

> ### Run Schematics by using the `schematic` and `externalSchematic` function

Reviewer: () behind fn names

> Perfect, we can now execute and combine our Schematics.

Reviewer: Consistent spelling of "schematics": upper or lowercase?

Author: Alsoways uppercase

Reviewer: okay! :-) wasn't sure

}
```

If we now run `schematics ../playground/src/collection.json:ng-add --debug=false` from our example Angular project, we can see that the `ng-add` schematic has called the `playground` schematic.

With this knowledge you can define small atomic Schematics that can be executed 'standalone' or from another Schematic that combines multiple standalone Schematics and calls them with specific parameters.

- [Check out the example in the playground repository on GitHub](https://github.com/d-koppenhagen/schematics-helpers-playground/tree/master/playground/src/ng-add)

### Run Schematics by using the `schematic()` and `externalSchematic()` function

Perfect, we can now execute and combine our Schematics.

> But what if we want to actually combine external Schematics developed by others and integrate them in our Schematic?

Reviewer: ```suggestion
But what if we want to combine external Schematics developed by others and integrate them in our own Schematics?
```

> As we are actually lazy we don't want to write in the docs that others have to execute some other things before and then run our Schematic.

Reviewer: ```suggestion
Users are lazy, so we don't want to leave it up to them to manually execute some other things before running our Schematics.
```

> Imagine you are working in a big company with lot's of Angular projects that are ready to start.

Reviewer: ```suggestion
Imagine you are working in a big company with multiple different Angular projects.
```

> This company has its own UI library already and all the applications are very different and ran by different teams (so not really a use case for a Monorepo).
> But there are also things they all have in common like a Single-Sing-On.
> Also, the basic design looks always similar (at least the header and the footer of the apps).
> 
> Often I've seen companies building a reference implementation for such apps that's cloned / copied and adjusted by all developers.
> But there are some problems with this kind of workflow:
> 
> - You have to keep the reference project always up-to-date
> - You have to clean up your copy of the project from stuff you don't need
> - You need to tell all teams copying this reference to keep track of changes and to adjust their copies frequently
> 
> So a better solution in my opinion is to use the Schematics.
> You can create a `ng new` schematic that will scaffold the whole project code for you.
> But you don't want to start from scratch.
> So you probably want to combine things like this:

Reviewer: ```suggestion
This company already has its own standardized UI library, but all the applications are very different and ran by different teams (so not really a use case for a Monorepo).
However, there are also things they all have in common like a Single Sign-On.
Also, the basic design always looks similar – at least the header and the footer of all the apps.

I've often seen companies building a reference implementation for such apps that's then cloned / copied and adjusted by all developers.
However, there are some problems with this kind of workflow:

- You always have to keep the reference project up-to-date.
- You have to clean up your copy of the project from stuff you don't need.
- You need to tell all teams to copy this reference to keep track of changes and to adjust their copies frequently.

Thus, a better solution in my opinion is to use Schematics for the whole integration and upgrade workflow.
You can create an `ng new` schematic that will scaffold the whole project code for you.
But you don't want to start from scratch, so you probably want to combine things like these:
```

> - `ng add`: Add your Schematic to existing projects
>   - Cooperate UI library (always)
>   - Single-Sign-On (optional)
>   - Header-Component (optional)
>   - Footer-Component (optional)
> - `ng new`: Create a new project with your company defaults
>   - Create the basic application generated from the Angular CLI (**`externalSchematic`**)
>   - Run the `ng add` schematic
> 
> Alright, we already know how we can achieve the most of the things.
> But one thing we haven't learned yet: How to run other (external) Schematics?
> Therefore, we can use the `externalSchematic` function.
> 
> But first things first, let's check if our collection file is ready to start:

Reviewer: ```suggestion
- `ng add`: Add your Schematics to an existing project
  - Corporate UI library (always)
  - Single Sign-On (optional)
  - Header Component (optional)
  - Footer Component (optional)
- `ng new`: Create a new project with your company defaults
  - Create the basic application generated by the Angular CLI (**`externalSchematic`**)
  - Run the `ng add` Schematic

Alright, we already know how we can achieve most of these things.
However, there's one thing we haven't learned yet: How to run other (external) Schematics?
We can use the `externalSchematic` function for this.

But first things first, let's check if our collection file is ready to start:
```

```json
{
  "$schema": "../node_modules/@angular-devkit/schematics/collection-schema.json",
  "schematics": {
    "ng-add": {

>       "description": "Call other Schematics from the same or other packages.",

Reviewer: ```suggestion
      "description": "Call other Schematics from the same or other packages",
```

      "factory": "./ng-add/index#playground"
    },
    "ng-new": {
      "description": "Execute `ng new` with predefined options and run other stuff",
      "factory": "./ng-new/index#playground"
    }
  }
}
```

> > Using the special Schematic namen `ng-add` and `ng new` let's you later use the schematic with executing `ng add`/`ng new` (instead of other Schematics called with `ng generate`).
> > There is also the special Schematic name `ng-update`, that will be called in the end with the `ng update` Angular CLI command.

Reviewer: ```suggestion
> Using the special Schematic names `ng-add` and `ng-new` let's you later use the schematic by just executing `ng add`/`ng new` (instead of other Schematics called with `ng generate`).
> There is also a special Schematic named `ng-update` which will be called in the end with the `ng update` Angular CLI command.
```

> Okay, now we can start implementing our Schematic.
> To execute an external Schematic, it must of course be available somewhere.
> But as we want to create a complete new project with your `ng new` Schematic, we don't have any `node_modules` installed in the target directory where we want to initialize the Angular workspace.
> We can use the [`spawn` method from `child_process`](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) (available globally for node.js).
> This creates a new process that executes a command (in our case: `npm install @schematics/angular`).
> So make things sync we wrap the method call into a Promise and `await`ing for the Promise to be resolved.
> Now we listen to the `close` event from `spawn` and checking there was no error during install (code equals `0`).
> If everything worked fine, we will resolve the Promise.
> Otherwise, we can throw an error.
> The last step is to `chain` the `Rule`s we have.
> First we can now use the `externalSchematic` function to run the `ng new` Schematic from Angular itself and set up the basic app.
> We will hand over here some default options we always want to be set (such a using `SCSS`, support legacy browsers, strict mode, etc.).
> The second call is our `ng add` Schematic that adds our company specific components, UI libs and so on to the project.
> 
> > We've already learned running a local Schematic by using the `RunSchematicTask` class that we add to our `context` object.
> > In this example we are using the `schematic` function to achieve the same goal.
> > Why are there two ways? To be honest: I actually don't know.
> > Checking the source code inside the Angular CLI, I found both implementations.
> 

Reviewer: ```suggestion
After we defined the schema, we can now start to implement our Schematics.
To execute an external Schematic, it must be available in the scope of the project..
However, since we want to create a completely new project with the `ng new` Schematic, we don't have any `node_modules` installed in the target directory where we want to initialize the Angular workspace.
To run an external command we can use the [`spawn` method from `child_process`](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) (available globally for Node.js).
This creates a new process that executes a command (in our case: `npm install @schematics/angular`).
To make things look synchronous we wrap the method call into a Promise and `await` for the Promise to be resolved.
Now we listen to the `close` event from `spawn` and check that there was no error during install (code equals `0`).
If everything worked fine, we will resolve the Promise, otherwise we can throw an error.
The last step is to `chain` all of our `Rule`s:
We first use the `externalSchematic()` function to run the `ng new` Schematic from Angular itself and set up the basic app.
We will hand over some default options here such a using `SCSS`, support legacy browsers, strict mode, etc.
The second call is our `ng add` Schematic that adds our company specific components, UI libs and so on to the project.

> We've already learned how to run a local Schematic by using the `RunSchematicTask` class that we need to add to our `context` object.
> In this example we are using the `schematic()` function to achieve the same goal.
> Why are there two ways? To be honest: I actually don't know.
> I found both implementations in the source code of Angular CLI.

```

```ts
import {
  Rule,
  SchematicContext,
  Tree,
  externalSchematic,
  schematic,
  chain
} from '@angular-devkit/schematics';
import {
  Schema as AngularNgNewSchema,
  PackageManager,
  Style
} from '@schematics/angular/ng-new/schema';
import { spawn } from 'child_process';

export function playground(options: AngularNgNewSchema): Rule {
  return async (_tree: Tree, _context: SchematicContext) => {
    const angularSchematicsPackage = '@schematics/angular';
    const ngNewOptions: AngularNgNewSchema = {

>       version: '10.1.0',

Reviewer: Why is the version fixed?

Author: Because the schematics of the defined version are executed. Probably I should write a sentence somewhere to clarify it.

Author: because the Angular Schematics needs this option as it will run the specific schematic version. Thats probably something special about the `ng new` Schematic from Angular.

Author: I've added a short sentence for explanation

      name: options.name,
      routing: true,
      strict: true,
      legacyBrowsers: true,
      style: Style.Scss,
      packageManager: PackageManager.Npm
    }
    await new Promise<boolean>((resolve) => {
      console.log('📦 Installing packages...');
      spawn('npm', ['install', angularSchematicsPackage])
        .on('close', (code: number) => {
          if (code === 0) {
            console.log('📦 Packages installed successfully ✅');
            resolve(true);
          } else {
            throw new Error(
              `❌ install Angular Schematics from '${angularSchematicsPackage}' failed`
            );
          }
        });
    });
    return chain([
      externalSchematic(angularSchematicsPackage, 'ng-new', ngNewOptions),
      schematic('ng-add', {})
    ]);
  };
}
```

> When we will run now the `ng new` Schematic from somewhere outside of an Angular workspace, we can see that first the Angular `ng new` Schematic is executed with our predefined settings.
> After this, the `ng add` Schematics is called.

Reviewer: ```suggestion
When we now run the `ng new` Schematic from somewhere outside of an Angular workspace, we can see that first of all, the Angular `ng new` Schematic is executed with our predefined settings.
After this, the `ng add` Schematics is called.
```

📦 Installing packages...
📦 Packages installed successfully ✅
? What name would you like to use for the new workspace and initial project? my-project
CREATE my-project/README.md (1027 bytes)
CREATE my-project/.editorconfig (274 bytes)
CREATE my-project/.gitignore (631 bytes)
CREATE my-project/angular.json (3812 bytes)
...
CREATE my-project/src/app/app.component.scss (0 bytes)
CREATE my-project/src/app/app.component.html (25757 bytes)
CREATE my-project/src/app/app.component.spec.ts (1069 bytes)
CREATE my-project/src/app/app.component.ts (215 bytes)
CREATE my-project/src/app/package.json (816 bytes)
CREATE my-project/e2e/protractor.conf.js (869 bytes)
CREATE my-project/e2e/tsconfig.json (294 bytes)
CREATE my-project/e2e/src/app.e2e-spec.ts (643 bytes)
CREATE my-project/e2e/src/app.po.ts (301 bytes)
⠏ Installing packages...
✔ Packages installed successfully.
schematic works
```

After you have deployed the Schematic, you can now execute it by running:

```bash
npm i -g my-schematic-package-name # install the Schematic so it's available globally
ng new my-app --collection=my-schematic-package-name # Run the Angular CLI's `ng new` Schematic with the defined collection
```

> Similar to this example you can call just the `ng add` Schematic from the collection if you are in an existing Angular workspace:

Reviewer: ```suggestion
Similar to this example you can call the `ng add` Schematic from the collection if you are in an existing Angular workspace:
```
